src/main/java/eu/dissco/core/digitalspecimenprocessor/service/HandleService.java

```diff
@@ -30,62 +64,261 @@
 @RequiredArgsConstructor
 public class HandleService {
 
+  private static final String HANDLE_PROXY = "https://hdl.handle.net/";
+  private static final String FDO_PROFILE_LOC = HANDLE_PROXY + "21.T11148/d8de0819e144e4096645";
+  private static final String DIGITAL_OBJECT_TYPE_LOC =
+      HANDLE_PROXY + "21.T11148/894b1e6cad57e921764e";
+  private static final byte[] DIGITAL_OBJECT_TYPE_VALUE = "digitalSpecimen".getBytes(
+      StandardCharsets.UTF_8);
+  private static final byte[] FDO_RECORD_LICENSE_LOC = "https://creativecommons.org/publicdomain/zero/1.0/".getBytes(
+      StandardCharsets.UTF_8);
+  private static final HashSet<String> notTypes = new HashSet<>(
+      Arrays.asList("false", "specimen", "type"));
+
+  private static final String ODS_PREFIX = "ods:";
+
+  private static final String UNKNOWN = "Unknown";
+
   private static final String PREFIX = "20.5000.1025/";
-  private static final String HANDLE = "Handle";
-  private static final String DIGITAL_OBJECT_SUBTYPE = "digitalObjectSubtype";
   private static final String SPECIMEN_HOST = "specimenHost";
   private static final String TO_BE_FIXED = "Needs to be fixed!";
-  private static final String DUMMY_HANDLE = "http://hdl.handle.net/21...";
   private final Random random;
   private final char[] symbols = "ABCDEFGHJKLMNPQRSTVWXYZ1234567890".toCharArray();
   private final char[] buffer = new char[11];
-  private final ObjectMapper mapper;
   private final DocumentBuilder documentBuilder;
   private final HandleRepository repository;
   private final TransformerFactory transformerFactory;
+  private final DateTimeFormatter dt = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS",
+      Locale.ENGLISH).withZone(ZoneId.of("UTC"));
 
   public String createNewHandle(DigitalSpecimen digitalSpecimen)
-      throws TransformerException {
+      throws TransformerException, PidCreationException {
+    checkForPrimarySpecimenObjectId(digitalSpecimen);
     var handle = generateHandle();
     var recordTimestamp = Instant.now();
-    var handleAttributes = fillPidRecord(digitalSpecimen, handle, recordTimestamp);
+    var handleAttributes = fillFdoRecord(digitalSpecimen, handle, recordTimestamp);
     repository.createHandle(handle, recordTimestamp, handleAttributes);
     return handle;
   }
 
-  private List<HandleAttribute> fillPidRecord(DigitalSpecimen digitalSpecimen, String handle,
-      Instant recordTimestamp)
+  private void checkForPrimarySpecimenObjectId(DigitalSpecimen digitalSpecimen)
+      throws PidCreationException {
+    var primarySpecimenObjectId = digitalSpecimen.physicalSpecimenId()
+        .getBytes(StandardCharsets.UTF_8);
+    var optionalRecord = repository.searchByPrimarySpecimenObjectId(primarySpecimenObjectId);
+    if (optionalRecord.isPresent()) {
+      throw new PidCreationException(
```

Comment:
I would say we update the PID record and attach it to the specimen instead of an exception. Might be something to discuss with Sharif as it involves the workflow.

Reply:
Good idea, changed